### ISSUE TITLE
Implement & test sigmoid loss

### ISSUE THREAD

**Issue by rwightman** [mlfoundations/open_clip#476]
a la: https://arxiv.org/abs/2303.15343

**Comment by lucidrains** [comment]
![Screenshot from 2023-03-29 13-35-17](https://user-images.githubusercontent.com/108653/228661201-9e7a5775-47e6-4639-a69d-a6e185e6dbaf.png)

why is there a negative in front of the `dot` ? don't we want the diagonal (matching text image pairs) to have a high logit, and correspondingly low -log_sigmoid? (label diagonal seems to be all positive ones)

**Comment by rwightman** [comment]
@lucidrains yeah, think it's a typo, I ran through a toy case and loss is -ve when features all match

**Comment by lucidrains** [comment]
yea I thought so; ok maybe I can knock this out Friday morning, if you don't get to it first

**Comment by rwightman** [comment]
@lucidrains I haven't written any code for the core loss at this point, I've been pondering the distributed bit because doing that efficiently will be key to making this stand out relative to InfoNCE in our training @ scale situations. I'm pretty sure the authors are using jax primitives that don't exist in pytorch so might have to manually imp functions that do send/recv pairs between node pairs, I think ie jax ppermute is a bit higher level for doing this sort of operation.

**Comment by lucidrains** [comment]
oh yeah you are right, this may not be straightforward

**Comment by lyprince** [comment]
You can use a custom ring reduce technique with `send`, `recv` pairs,  which will work for infoNCE too.

e.g., On CPU:

```
import os
import torch
import torch.distributed as dist
import torch.multiprocessing as mp


def ring_infonce(x_embs, y_embs, logit_scale):
    y_embs_send_buff = y_embs.clone()
    y_embs_recv_buff = y_embs.clone()

    local_logits = logit_scale * x_embs @ y_embs.T
    pos_pair_loss = -local_logits.diag()
    neg_pair_loss = torch.logsumexp(local_logits, dim=1, keepdim=True)

    size = dist.get_world_size()
    rank = dist.get_rank()
    left = ((rank - 1) + size) % size
    right = (rank + 1) % size
    for t in range(size - 1):
        if t % 2 == 0:
            # Send send_buff
            send_req = dist.isend(y_embs_send_buff, right)
            dist.recv(y_embs_recv_buff, left)
            cross_logits = logit_scale * x_embs @ y_embs_recv_buff.T
            neg_pair_loss[:] = torch.logsumexp(
                torch.cat((neg_pair_loss, cross_logits), dim=1), dim=1, keepdim=True
            )
        else:
            send_req = dist.isend(y_embs_recv_buff, right)
            dist.recv(y_embs_send_buff, left)
            cross_logits = logit_scale * x_embs @ y_embs_send_buff.T
            neg_pair_loss[:] = torch.logsumexp(
                torch.cat((neg_pair_loss, cross_logits), dim=1), dim=1, keepdim=True
            )
        send_req.wait()
    loss = pos_pair_loss + neg_pair_loss.squeeze()
    return loss.mean()


def run(rank, size):
    setup(rank=rank, world_size=size)
    torch.manual_seed(100 + rank)
    logit_scale = 1.0
    x_embs = torch.randn(2, 2)
    y_embs = torch.randn(2, 2)
    print(rank, y_embs)
    loss = ring_infonce(x_embs, y_embs, logit_scale)
    print(f"Rank: {rank}, loss: {loss}", flush=True)
    cleanup()


def setup(rank, world_size):
    """Initialize the distributed environment."""
    os.environ["MASTER_ADDR"] = "127.0.0.1"
    os.environ["MASTER_PORT"] = "29500"
    dist.init_process_group("gloo", rank=rank, world_size=world_size)


def cleanup():
    dist.destroy_process_group()


if __name__ == "__main__":
    size = 4
    print("With distributed")

    mp.spawn(run, args=(size,), nprocs=size, join=True)

    print("Without distributed")
    x_embs = []
    y_embs = []
    logit_scale = 1.0
    for i in range(size):
        torch.manual_seed(100 + i)
        x_embs.append(torch.randn(2, 2))
        y_embs.append(torch.randn(2, 2))
    x_embs = torch.cat(x_embs)
    y_embs = torch.cat(y_embs)
    logits = logit_scale * x_embs @ y_embs.T
    pos_pair_loss = -logits.diag()
    neg_pair_loss = torch.logsumexp(logits, dim=1)
    loss = pos_pair_loss + neg_pair_loss
    losses = [l_.mean() for l_ in torch.split(loss, 2)]
    for ix, l_ in enumerate(losses):
        print(f"Rank {ix}: loss: {l_}")

```

**Comment by lyprince** [comment]
Apologies for the opinionated formatting: 

https://github.com/mlfoundations/open_clip/compare/main...lyprince:open_clip:ring-infonce?diff=split#diff-8a62ee1fa8d88ce27d80f8678ebf59bd908b6f7d357452def316d6554363d9b1R80
